### PR TITLE
Implemented BOOST_PFR_ENABLED macro

### PR DESCRIPTION
**Pull request by denzor200**
New macro tested on the Godbolt in so many variants:
case 1 (before) : https://godbolt.org/z/hjhKf3h5n
case 1 (after) : https://godbolt.org/z/vfeKfcMh7
case 2 (before) : https://godbolt.org/z/bvfocY1v1
case 2 (after) : https://godbolt.org/z/bnYboGYa5
case 3 (before) : https://godbolt.org/z/es7cEoo75
case 3 (after) : https://godbolt.org/z/sMdr1vGPW

### CODE REVIEW

**Comment by denzor200** [comment]
I think this is a bad idea. But we will not close it yet

File: include/boost/pfr/config.hpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+#   if __has_cpp_attribute(maybe_unused)
+#       define BOOST_PFR_MAYBE_UNUSED [[maybe_unused]]
+#   endif
+#endif
+
+#ifndef BOOST_PFR_MAYBE_UNUSED
+#   define BOOST_PFR_MAYBE_UNUSED
+#endif
+
+#ifndef BOOST_NO_PFR
```
Comment (denzor200) [inline include/boost/pfr/config.hpp:94]: Add `BOOST_NO_PFR` macro into:
-doc/pfr.qbk (near another macroses)
-test/print_config.cpp (and change inclusion core.hpp -> config.hpp)

**Comment by denzor200** [comment]
> I think this is a bad idea.. But we will not close it yet

This idea may become good if we provide the way to override compiler's decision manually. Like this:
```
#define BOOST_NO_PFR 1    //< force set to 'disallowed'
#include <boost/pfr.hpp>  //< compile time error, PFR not supported by user's decision
```
```
#define BOOST_NO_PFR 0    //< force set to 'allowed'
#include <boost/pfr.hpp>  //< good when your compiler really supports pfr
```
```
// #define BOOST_NO_PFR 0  //< let compiler set PFR's status
#include <boost/pfr.hpp>   //< good when your compiler support pfr
```

I'll continue working on it.

File: include/boost/pfr/detail/config.hpp [hunk 2]
```diff
@@ -7,85 +7,18 @@
...
-#       define BOOST_PFR_MAYBE_UNUSED [[maybe_unused]]
-#   endif
-#endif
-
-#ifndef BOOST_PFR_MAYBE_UNUSED
-#   define BOOST_PFR_MAYBE_UNUSED
-#endif
+#error Boost.PFR library is not supported in your environment.             \
+       Try one of the possible solutions:                                  \
+       1. enable C++14;                                                    \
```
Comment (denzor200) [inline include/boost/pfr/detail/config.hpp:16]: First, we should propose user to take away an `-DBOOST_NO_PFR=1`, if it exist.
Then all another.

File: doc/pfr.qbk [hunk 3]
```diff
@@ -469,6 +469,7 @@ By default Boost.PFR [*auto-detects your compiler abilities] and automatically d
     [[*BOOST_PFR_USE_STD_MAKE_INTEGRAL_SEQUENCE*] [Define to `0` if you are hit by the template instantiation depth issues with `std::make_integer_sequence` and wish to use Boost.PFR version of that metafunction. Define to `1` to override Boost.PFR detection logic. ]]
     [[*BOOST_PFR_HAS_GUARANTEED_COPY_ELISION*] [Define to `0` if your compiler does not implement C++17 guaranteed copy elision properly and fails to reflect aggregates with non-movable fields. Define to `1` to override Boost.PFR detection logic. ]]
     [[*BOOST_PFR_ENABLE_IMPLICIT_REFLECTION*] [Define to `0` if you are hit by lots of non-effective choices made by implicitly reflection. Define to `1` to override Boost.PFR detection logic. ]]
+    [[*BOOST_NO_PFR*] [On platforms where Boost.PFR is not supported, the boost/pfr/config.hpp header defines the BOOST_NO_PFR macro equals 1. Defining this macro as 1 before including the header disables the Boost.PFR library. ]]
```
Comment (apolukhin) [inline doc/pfr.qbk:472]:
```suggestion
    [[*BOOST_NO_PFR*] [On platforms where Boost.PFR is not supported, the `boost/pfr/config.hpp` header defines the BOOST_NO_PFR macro equal to 1. Defining this macro as 1 before including the header disables the Boost.PFR library. ]]
```

Comment (apolukhin) [inline doc/pfr.qbk:472]: Please change the file permossin back to non-executable

File: include/boost/pfr/detail/config.hpp [hunk 4]
```diff
@@ -7,94 +8,19 @@
 #define BOOST_PFR_DETAIL_CONFIG_HPP
 #pragma once
 
-#include <type_traits> // to get non standard platform macro definitions (__GLIBCXX__ for example)
+#include <boost/pfr/config.hpp>
```
Comment (apolukhin) [inline include/boost/pfr/detail/config.hpp:11]: Please change the file permission back to non-executable

File: include/boost/pfr/detail/config.hpp [hunk 5]
```diff
@@ -7,94 +8,19 @@
...
 
-// Reminder:
-//  * MSVC++ 14.2 _MSC_VER == 1927 <- Loophole is known to work (Visual Studio ????)
-//  * MSVC++ 14.1 _MSC_VER == 1916 <- Loophole is known to NOT work (Visual Studio 2017)
-//  * MSVC++ 14.0 _MSC_VER == 1900 (Visual Studio 2015)
-//  * MSVC++ 12.0 _MSC_VER == 1800 (Visual Studio 2013)
+#if BOOST_NO_PFR
```
Comment (apolukhin) [inline include/boost/pfr/detail/config.hpp:13]: Negation in macro is not nice. How about changing the macro name to `BOOST_PFR_ENABLED`?

File: test/print_config.cpp [hunk 6]
```diff
@@ -3,7 +3,7 @@
 // Distributed under the Boost Software License, Version 1.0. (See accompanying
 // file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 
-#include <boost/pfr/core.hpp>
```
Comment (apolukhin) [inline test/print_config.cpp:6]: chmod -x

**Review by apolukhin — CHANGES_REQUESTED** [review]

**Review by apolukhin — APPROVED** [review]

**Pull request merged**